File: test/mock_registry.py
"""
Minimal OCI registry mock for resolve.sh tests: bearer auth, paginated tag list, digests per tag.
"""
import base64, json, sys
from http.server import BaseHTTPRequestHandler, HTTPServer
from urllib.parse import urlparse, parse_qs

PORT = int(sys.argv[1])
DIGESTS = {"stable": "sha256:aaa", "0.61.0": "sha256:aaa", "0.62.0": "sha256:bbb", "0.60.0": "sha256:ccc", "0.59.0": "sha256:ddd", "0.60.0-alpha.3": "sha256:eee", "0.62.0-rc.1": "sha256:fff"}
PAGE1 = ["0.59.0", "0.60.0-alpha.3", "0.60.0"]
PAGE2 = ["0.61.0", "0.62.0-rc.1", "0.62.0", "stable"]


class H(BaseHTTPRequestHandler):
    """
    Serves the handful of registry endpoints resolve.sh touches.
    """

    def log_message(self, *a):
        pass

    def do_GET(self):
        u = urlparse(self.path)
        q = parse_qs(u.query)
        if u.path == "/v2/":
            self.send_response(401)
            self.send_header("Www-Authenticate", f'Bearer realm="http://127.0.0.1:{PORT}/service/token",service="harbor-registry"')
            self.end_headers(); return
        if u.path == "/service/token":
            auth = self.headers.get("Authorization", "")
            ok = auth.startswith("Basic ") and base64.b64decode(auth[6:]).decode() == "robot$abc:s3cret"
            if not ok:
                self.send_response(401); self.end_headers(); return
            body = json.dumps({"token": "tok123"}).encode()
            self.send_response(200); self.send_header("Content-Type", "application/json"); self.end_headers(); self.wfile.write(body); return
        if self.headers.get("Authorization") != "Bearer tok123":
            self.send_response(401); self.end_headers(); return
        if u.path == "/v2/valley/valley/tags/list":
            if "last" in q:
                body = json.dumps({"name": "valley/valley", "tags": PAGE2}).encode()
                self.send_response(200); self.end_headers(); self.wfile.write(body); return
            body = json.dumps({"name": "valley/valley", "tags": PAGE1}).encode()
            self.send_response(200); self.send_header("Link", '</v2/valley/valley/tags/list?n=100&last=0.60.0>; rel="next"'); self.end_headers(); self.wfile.write(body); return
        if u.path.startswith("/v2/valley/valley/manifests/"):
            tag = u.path.rsplit("/", 1)[1]
            if tag in DIGESTS:
                self.send_response(200); self.send_header("Docker-Content-Digest", DIGESTS[tag]); self.end_headers(); return
            self.send_response(404); self.end_headers(); return
        if u.path.startswith("/v2/valley/valley-alpha/"):
            self.send_response(404); self.end_headers(); return
        self.send_response(404); self.end_headers()


HTTPServer(("127.0.0.1", PORT), H).serve_forever()

File: test/mock-registry.js
import { createHash } from "node:crypto";
import { createServer } from "node:http";
import { gzipSync } from "node:zlib";

const TOKEN = "tok123";
export const MOCK_USER = "robot$abc";
export const MOCK_PASSWORD = "s3cret";
export const CHART_BYTES = gzipSync(Buffer.from("fake chart tarball"));
const HELM_CHART_LAYER = "application/vnd.cncf.helm.chart.content.v1.tar+gzip";

/**
 * Returns the sha256 digest of a buffer in OCI form.
 */
function sha256Digest(buffer) {
  return `sha256:${createHash("sha256").update(buffer).digest("hex")}`;
}

const CHART_DIGEST = sha256Digest(CHART_BYTES);

/**
 * Builds a manifest body for one artifact; `artifact` makes each body, and so each digest, distinct.
 */
function manifestBody(artifact, layerDigest = CHART_DIGEST) {
  return Buffer.from(JSON.stringify({ schemaVersion: 2, annotations: { artifact }, layers: [{ mediaType: HELM_CHART_LAYER, digest: layerDigest, size: CHART_BYTES.length }] }));
}

// Tags sharing an artifact share a digest; "stable" and 0.61.0 are one artifact so the resolver must pick 0.61.0, not the highest tag
const ARTIFACTS = {
  stable: "a",
  "0.61.0": "a",
  "0.62.0": "b",
  "0.60.0": "c",
  "0.59.0": "d",
  "0.60.0-alpha.3": "e",
  "0.62.0-rc.1": "f",
};
const PAGE1 = ["0.59.0", "0.60.0-alpha.3", "0.60.0"];
const PAGE2 = ["0.61.0", "0.62.0-rc.1", "0.62.0", "stable"];

// Tampered artifacts, reachable only by name: a lying digest header, and a layer digest the blob does not match
const BAD_HEADER = "bad-header";
const BAD_BLOB = "bad-blob";
const BAD_BLOB_DIGEST = sha256Digest(Buffer.from("some other blob"));

/**
 * Sends a JSON body with the given status and extra headers.
 */
function json(res, status, body, headers = {}) {
  res.writeHead(status, { "content-type": "application/json", ...headers });
  res.end(JSON.stringify(body));
}

/**
 * Starts a mock OCI registry that mimics Harbor's bearer auth, paginated tag list, manifests, and one chart blob. Returns { port, close }.
 */
export async function startMockRegistry() {
  const server = createServer((req, res) => {
    const url = new URL(req.url, "http://127.0.0.1");
    const port = server.address().port;

    // Auth challenge and token exchange
    if (url.pathname === "/v2/") {
      res.writeHead(401, { "www-authenticate": `Bearer realm="http://127.0.0.1:${port}/service/token",service="harbor-registry"` });
      return res.end();
    }
    if (url.pathname === "/service/token") {
      const auth = req.headers.authorization ?? "";
      const creds = auth.startsWith("Basic ") ? Buffer.from(auth.slice(6), "base64").toString() : "";
      if (creds !== `${MOCK_USER}:${MOCK_PASSWORD}`) return json(res, 401, { errors: [{ code: "UNAUTHORIZED" }] });
      return json(res, 200, { token: TOKEN });
    }
    if (req.headers.authorization !== `Bearer ${TOKEN}`) return json(res, 401, { errors: [{ code: "UNAUTHORIZED" }] });

    // Tag list in two pages
    if (url.pathname === "/v2/valley/valley/tags/list") {
      if (url.searchParams.has("last")) return json(res, 200, { name: "valley/valley", tags: PAGE2 });
      return json(res, 200, { name: "valley/valley", tags: PAGE1 }, { link: `</v2/valley/valley/tags/list?n=100&last=0.60.0>; rel="next"` });
    }

    // Manifests carry the digest header and a Helm chart layer
    const manifest = /^\/v2\/valley\/valley\/manifests\/(.+)$/.exec(url.pathname);
    if (manifest) {
      const tag = manifest[1];
      let body;
      let digest;
      if (tag === BAD_HEADER) {
        body = manifestBody("tampered");
        digest = sha256Digest(manifestBody("a"));
      } else if (tag === BAD_BLOB) {
        body = manifestBody("bad-blob", BAD_BLOB_DIGEST);
        digest = sha256Digest(body);
      } else if (ARTIFACTS[tag]) {
        body = manifestBody(ARTIFACTS[tag]);
        digest = sha256Digest(body);
      } else {
        return json(res, 404, { errors: [{ code: "MANIFEST_UNKNOWN" }] });
      }
      res.writeHead(200, { "content-type": "application/vnd.oci.image.manifest.v1+json", "docker-content-digest": digest });
      return res.end(body);
    }

    // Both blob digests serve the same bytes, so only CHART_DIGEST matches what arrives
    if (url.pathname === `/v2/valley/valley/blobs/${CHART_DIGEST}` || url.pathname === `/v2/valley/valley/blobs/${BAD_BLOB_DIGEST}`) {
      res.writeHead(200, { "content-type": "application/octet-stream" });
      return res.end(CHART_BYTES);
    }

    // The alpha repository has nothing promoted
    return json(res, 404, { errors: [{ code: "NOT_FOUND" }] });
  });

  await new Promise((resolve) => server.listen(0, "127.0.0.1", resolve));
  return { port: server.address().port, close: () => new Promise((resolve) => server.close(resolve)) };
}

export { BAD_BLOB, BAD_HEADER, sha256Digest };
